cudax/include/cuda/experimental/__container/vector.cuh

```diff
@@ -0,0 +1,1573 @@
+  //! @param __dest Pointer to the start of the output segment.
+  //! @returns Pointer equal to `__dest + __count`
+  static pointer __copy_same(const_pointer __first, const_pointer __last, pointer __dest)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::copy(__first, __last, __dest);
+    }
+    else
+    {
+      const auto __count = static_cast<size_t>(__last - __first);
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemcpy, "failed to copy data", __dest, __first, sizeof(_Tp) * __count, ::cudaMemcpyDeviceToDevice);
+      return __dest + __count;
+    }
+    _CCCL_UNREACHABLE();
+  }
+
+  //! @brief Copies \p __count elements from `[__first, __last)` to \p __dest, where \p __first and \p __dest reside in
+  //! the different memory spaces
+  //! @param __first Pointer to the start of the input segment.
+  //! @param __last Pointer to the end of the input segment.
+  //! @param __dest Pointer to the start of the output segment.
+  //! @param __count The number of elements to be copied.
+  template <class _Iter, cudaMemcpyKind __kind = __is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice>
+  static void __copy_cross(_Iter __first, _Iter __last, pointer __dest, size_type __count)
+  {
+    _CCCL_IF_CONSTEXPR (__kind == cudaMemcpyHostToHost)
+    {
+      _CUDA_VSTD::copy(__first, __last, __dest);
+    }
+    _CCCL_ELSE_IF_CONSTEXPR (!_CUDA_VSTD::contiguous_iterator<_Iter>)
+    { // For non-coniguous iterators we need to copy into temporary host storage to use cudaMemcpy
+      // This should only ever happen when passing in data from host to device
+      _CCCL_ASSERT(__kind == cudaMemcpyHostToDevice, "Invalid use case!");
+      auto __temp = _CUDA_VSTD::get_temporary_buffer<_Tp>(__count).first;
+      _CUDA_VSTD::copy(__first, __last, __temp);
+      _CCCL_TRY_CUDA_API(::cudaMemcpy, "failed to copy data", __dest, __temp, sizeof(_Tp) * __count, __kind);
+      _CUDA_VSTD::return_temporary_buffer(__temp);
+    }
+    else
+    {
+      (void) __last;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemcpy, "failed to copy data", __dest, _CUDA_VSTD::to_address(__first), sizeof(_Tp) * __count, __kind);
+    }
+  }
+
+  //! @brief Creates a gap in the current memory of size \p __count at position \p __pos
+  //! @param __pos Pointer to the first element to be moved.
+  //! @param __count The size of the empty space.
+  //! Depending on the available memory this might need to reallocate. Furthermore, if this executes on device then we
+  //! need temporary storage because we want to work in parallel.
+  void __create_gap(size_type __pos, size_type __count)
+  {
+    pointer __first            = __unwrapped_begin();
+    pointer __end              = __unwrapped_end();
+    pointer __middle           = __first + __pos;
+    const size_type __new_size = __size_ + __count;
+    if (capacity() - __size_ < __count)
+    { // need to reallocate
+      __buffer __new_buf{__borrow_resource(), __new_size};
```

Comment:
We are oversynchronizing here, once when allocating, once when calling thrust, once when destructiong the temp buffer

We should have a stream inside